### ISSUE TITLE
24bit files distortions problem at raspbian armv7l GNU/Linux if the internal player volume control is set to any value less than +0dB

### ISSUE THREAD

**Issue by eshchenko** [DeaDBeeF-Player/deadbeef#1978]
Dear Developers

I do understand, that my problem is rather exotic, but there is a big underestimate of the deadbeef
beauty in the Raspberry Pi community.
Meanwhile, recently they finally made the snd_usb_audio module for raspbian compatible with the 24bit.
Also, last alsa utilities programs and libraries work as expected (24bit is supported)... 
So, I hope it is time to support the Raspberry....
 
### Steps to reproduce the problem
Connected external usb card to Raspberry PI2:
DR DAC Prime:
pi@raspberrypi:~/development/alsacap $ ./alsacap
*** Scanning for playback devices ***
Card 0, ID `ALSA', name `bcm2835 ALSA'
......
**Card 1, ID `prime', name `Dr. DAC prime'
Device 0, ID `USB Audio', name `USB Audio', 1 subdevices (1 available)
2 channels, sampling rate 8000..96000 Hz
Sample formats: S16_LE, S24_3LE
Subdevice 0, name `subdevice #0'**
Device 1, ID `USB Audio', name `USB Audio #1', 1 subdevices (1 available)
2 channels, sampling rate 48000..48000 Hz
Sample formats: S16_LE
Subdevice 0, name `subdevice #0'

Compile DeadBeaf from the source.
Be sure the external usb card is used the most direct way:
DeadBeeF configuration: 
ALSA output plugin
Output device: Dr DAC Direct Hardware Device without any conversions
ALSA resampling is not checked 

Take a 24bit file, be sure that it can be played correctly by the aplay:
pi@raspberrypi:~/development/bin $ **aplay -D hw:CARD=prime,DEV=0 tmp.wav
Playing WAVE 'tmp.wav' : Signed 24 bit Little Endian in 3bytes, Rate 48000 Hz, Stereo**

Start to play this 24bit pcm (*.wav) file by the deadbeef player.

Cross-check that the hardware and alsa module work as expected e.g by inspecting
/proc/asound/ tree

### What's going on? Describe the problem in as much detail as possible
**Huge distortion, 
but only if the internal player volume control is set to any value less than +0dB**

If the volume is at maximum, the sound is clean without any distortion.

Temporary solution: 
Similar to suggested in
https://github.com/DeaDBeeF-Player/deadbeef/issues/1908
1) set volume to maximum
2) remove the volume control widget 

The same card connected to the PC ( x86_64 GNU/Linux ) plays the 24bit without any distortion 
independent on the volume control value (e.g. -6db),
so the problem reported is most probably related with the arm architecture.

Recommendation for the future - 
1) be sure that the default value of the playback.volume is 0.0 like mine: 
pi@raspberrypi:~/development/bin $ grep -a playback.volume ~/.config/deadbeef/config
playback.volume 0.0000000
2) do not use playback.volume

### Information about the software:

Deadbeef version:
Development, also tested at 
0.7.2
0.7.0
etc

OS:
Paspberri Pi 2B
Raspbian GNU/Linux 8 (jessie)
Linux raspberrypi 4.9.35-v7+ #1014 SMP Fri Jun 30 14:47:43 BST 2017 armv7l GNU/Linux

For me, the problem described is rather a feature and no action is needed,
but who knows, may be some person would like to use separate volume control.

Best regards, 
Dmitry Eshchenko

p.s. It is interesting to note that mplayer works fine even with the volume option:

mplayer -volume 50 -ao alsa:device=hw=1.0 tmp.wav
mplayer -af volume=-3.0:0 -ao alsa:device=hw=1.0 tmp.wav

**Comment by Alexey-Yakovenko** [comment]
Which version of the player? I'm pretty sure I fixed just this problem a few months ago

**Comment by eshchenko** [comment]
Dear Developers

*The problem is independent on the deadbeef version.*
Today I rechecked this by the development code which I got via
git clone
(before I used "download as zip")

I suspect it is the ARM processor architecture feature as the same card,
the same file and the same version of deadbeef works as accepted on the
Intel but not on the Raspberry Pi.

Also, I forgot to mention that at Rasbian
16bit works fine, the problem is with 24bit and volume control set to any
value less than +0 dB only.
The distortion becomes more and more pronounced while decreasing the
playback volume.

Actually this finding cost me a lot of time to understand as I used the gui
slider (mouse) to change the volume.
With this slider there is a feature - to be sure that the volume is at
maximum one has to slide the mouse to the very-very right even outside the
indicator.
Another problem is that the volume value in gui is shown in rounded - the
real slightly negative volume is reported as "0", and, as the distortions
are volume dependent, one can hear small distortion when the level is
reported (in realty - rounded) as "0".

As I already mentioned -
the workaround is to set the volume in ~/.config/deadbeef/config file to
zero:

playback.volume 0.0000000

and then remove the volume control from the gui.

As for your comment that you already corrected this problem few month ago -
have you tested this on the real ARM hardware and on the fresh Raspbian?
Please have in mind that all these features (24bit support by the
snd_usb_audio module and the correct support of 24bit by alsa utils and
libs) were introduced say  ~ 0.5 year ago.

Before all the attempts to play 24bit music by external usb card using alsa
were finished by the "automatic" downgrade into the 16bit mode.
That is why the content of /proc/asound is important.
_________________

*Bottom line:*

*Yes, the problem is still present in development 27.02.2017 version of the
player*
*It is present only on Linux raspberrypi 4.9.35-v7+ #1014 SMP Fri Jun 30
14:47:43 BST 2017 armv7l GNU/Linux*
_________________

Best regards,
Dmitry

**Comment by Alexey-Yakovenko** [comment]
> As for your comment that you already corrected this problem few month ago -
have you tested this on the real ARM hardware and on the fresh Raspbian?

Yes, but without Raspbian, was fixing this for android.

**Comment by Alexey-Yakovenko** [comment]
Please try this patch
```
--- a/streamer.c
+++ b/streamer.c
@@ -1815,7 +1815,7 @@ streamer_apply_soft_volume (char *bytes, int sz) {
             if (ivolume != 1000) {
                 int third = bytesread/3;
                 for (int i = 0; i < third; i++) {
-                    int32_t sample = ((unsigned char)stream[0]) | ((unsigned char)stream[1]<<8) | (stream[2]<<16);
+                    int32_t sample = ((unsigned char)stream[0]) | ((unsigned char)stream[1]<<8) | ((signed char)(stream[2]<<16));
                     int32_t newsample = (int32_t)((int64_t)sample * ivolume / 1000);
                     stream[0] = (newsample&0x0000ff);
                     stream[1] = (newsample&0x00ff00)>>8;
```

**Comment by eshchenko** [comment]
After patching and recompiling the behavior changed:
0 dB is still OK, but if I try to make the volume lower - the sound
disappears completely (no music) independent on how much I lower the volume.

Please also note the the before patching the distortions were observed in
two situations:
1) volume is less than 0dB
2) equalizer is enabled

After patching the equalizer does not change the spectrum - it seems the
equalizer is ignored

**Comment by Alexey-Yakovenko** [comment]
oh .. I actually made a typo..

should be this instead

```
--- a/streamer.c
+++ b/streamer.c
@@ -1815,7 +1815,7 @@ streamer_apply_soft_volume (char *bytes, int sz) {
             if (ivolume != 1000) {
                 int third = bytesread/3;
                 for (int i = 0; i < third; i++) {
-                    int32_t sample = ((unsigned char)stream[0]) | ((unsigned char)stream[1]<<8) | (stream[2]<<16);
+                    int32_t sample = ((unsigned char)stream[0]) | ((unsigned char)stream[1]<<8) | (((signed char)stream[2])<<16);
                     int32_t newsample = (int32_t)((int64_t)sample * ivolume / 1000);
                     stream[0] = (newsample&0x0000ff);
                     stream[1] = (newsample&0x00ff00)>>8;
```

**Comment by eshchenko** [comment]
Good news!
Now distortions disappeared - one can lower volume.
So the idea is correct! Thanks a lot from the Raspberry community!

What is still strange -
equalizer is not working... (behaves like disabled independent on the
enable/disable setting)
So I can not say that it is distortions free.
As I suspect, one has to propagate this patch into the equalizer...

Had I missed something in configuration?

The SuperEQ plugin is compiled and loaded.
No error messages regarding this plugin.





On Sat, Mar 10, 2018 at 10:38 AM, Alexey Yakovenko <notifications@github.com

**Comment by Alexey-Yakovenko** [comment]
FYI: it doesn't seem like equalizer issue is related to volume control in any way.
There was another issue with EQ on trunk which might have caused it. I'll verify.

**Comment by eshchenko** [comment]
Thanks a lot!!!
As for the EQ:

Yes, the EQ and volume distortions are independent.

As *EQ does not work (neither for 16b nor for 24bit)* *in the development
version*,
today I took the stable version 0.7.2, changed in the
streamer.c
the string

-                    int32_t sample = ((unsigned char)stream[0]) |
((unsigned char)stream[1]<<8) | (stream[2]<<16);
+                    int32_t sample = ((unsigned char)stream[0]) |
((unsigned char)stream[1]<<8) | (((signed char)stream[2])<<16);


and compiled this slightly modified version (on Raspberry).
As expected, the volume works fine (no distortions) but, as it was also
already expected, the* EQ produces distortions when playing 24bit only* (EQ
on 16bit is fine)

For the crosscheck I just did
git clone
and compiled the development version of the deadbeef on
Linux HP-EliteBook-8530p 4.4.0-040400-generic #201601101930 SMP Mon Jan 11
00:32:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
When starting the player from the terminal and playing a 24 bit file there
is a warning:



*cannot set sample format to 24 bps (error: Invalid argument), trying all
supported formatsFound compatible format 32 bps*

*It is interesting that there is no such warning at armv7l GNU/Linux - just
nothing!!!*

*Does it mean on Intel there is an additional file conversion???*

The equalizer on Intel works as expected (enable/disable works etc)

So, I can conclude that the changes between 0.7.2 and the current
development version result to *bypassing the EQ but on* the
Linux raspberrypi 4.9.35-v7+ #1014 SMP Fri Jun 30 14:47:43 BST 2017 *armv7l*
GNU/Linux
*only.*

So this EQ behavior is platform dependent.

Hope my notes can help you...

Should I make a new issue?


On Sun, Mar 11, 2018 at 10:14 AM, Alexey Yakovenko <notifications@github.com

**Comment by Alexey-Yakovenko** [comment]
Yes, please make a new issue about the EQ.
